File: agent-world-lab/backend/python/app/infrastructure/services/tool_execution_service.py
from __future__ import annotations

from app.adapters.abstractions import IProviderToolAdapter, ProviderToolContext
from app.application.interfaces import IToolExecutionService


class ToolExecutionService(IToolExecutionService):
    def __init__(self, adapters: list[IProviderToolAdapter]) -> None:
        self._adapters = adapters

    def execute(self, tool_name: str, tool_input: str | None, requested_by: str, environment: str | None, correlation_id: str) -> str:
        for adapter in self._adapters:
            if adapter.can_handle(tool_name):
                response = adapter.execute(
                    ProviderToolContext(
                        tool_name=tool_name,
                        tool_input=tool_input,
                        requested_by=requested_by,
                        environment=environment,
                        correlation_id=correlation_id,
                    )
                )
                return response.message
        raise ValueError(f"No adapter can handle tool '{tool_name}'")

File: agent-world-lab/backend/src/AgentWorldLab.Infrastructure/Services/ToolExecutionService.cs
using AgentWorldLab.Adapters.Abstractions;
using AgentWorldLab.Application.Abstractions;
using AgentWorldLab.Application.Models;
using AgentWorldLab.Domain.Models;

namespace AgentWorldLab.Infrastructure.Services;

public sealed class ToolExecutionService : IToolExecutionService
{
    private readonly IReadOnlyCollection<IProviderToolAdapter> _adapters;

    public ToolExecutionService(IEnumerable<IProviderToolAdapter> adapters)
    {
        _adapters = adapters.ToArray();
    }

    public ToolExecutionResult Execute(ToolDefinition tool, string? input, string requestedBy, string environment, string correlationId)
    {
        var adapter = _adapters.FirstOrDefault(a => a.CanHandle(tool.Name));
        if (adapter is null)
        {
            return new ToolExecutionResult("Failed", $"No provider adapter found for tool '{tool.Name}'.");
        }

        var context = new ProviderToolContext(tool.Name, input, correlationId, requestedBy, environment);
        var result = adapter.Execute(context);

        return new ToolExecutionResult(result.Status, $"[{result.Provider}] {result.Message}");
    }
}
